File: src/main/scala/com/digitalasset/testing/ledger/SandboxRunner.scala
/*
 * Copyright 2019 Digital Asset (Switzerland) GmbH and/or its affiliates
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package com.digitalasset.testing.ledger

import java.io.{File, IOException}
import java.util.Optional
import java.util.concurrent.TimeoutException
import java.util.function.Consumer

import com.daml.ledger.rxjava.DamlLedgerClient
import com.digitalasset.testing.ledger.clock.SandboxTimeProviderFactory
import com.digitalasset.testing.store.DefaultValueStore
import grizzled.slf4j.Logging

import scala.util.Try

/**
  * A class that can run a DAML Sandbox instance. A ledger client and a ledger adapter are initialized as well.
  * - To start, initialize the client and the adapter, call startSandbox.
  * - To stop, shutdown the client and the adapter, call stopSandbox.
  */
class SandboxRunner(val relativeDarPath: String,
                    val testModule: String,
                    val testScenario: String,
                    val sandboxPort: Int = 6865,
                    val waitTimeout: Int = 20,
                    val parties: Array[String] = Array.empty,
                    val setupApplication: Consumer[DamlLedgerClient])
    extends Logging {

  val damlLedgerClient: DamlLedgerClient =
    DamlLedgerClient.forHostWithLedgerIdDiscovery("localhost",
                                                  sandboxPort,
                                                  Optional.empty())
  val ledgerAdapter: DefaultLedgerAdapter =
    new DefaultLedgerAdapter(new DefaultValueStore(),
                             sandboxHost = "localhost",
                             sandboxPort = sandboxPort,
                             timeProviderFactory = SandboxTimeProviderFactory)
  private var sandbox: Process = null

  private def getSandboxRunner(scenario: String) =
    new ProcessBuilder("daml",
                       "sandbox",
                       "--",
                       "-p",
                       sandboxPort.toString,
                       "--scenario",
                       String.format("%s:%s", testModule, scenario),
                       relativeDarPath)
      .redirectError(new File("integration-test-sandbox.log"))
      .redirectOutput(new File("integration-test-sandbox.log"))

  def waitForConnection(damlLedgerClient: DamlLedgerClient,
                        seconds: Int): Unit = {
    var connected = false
    var time = seconds
    while ({
      !connected && 0 <= time
    }) try {
      damlLedgerClient.connect()
      connected = true
    } catch {
      case _ignored: Exception =>
        logger.info(s"Connecting to sandbox at localhost:${sandboxPort}")
        Try(Thread.sleep(1000))
    }
    if (!connected) {
      throw new TimeoutException(
        s"Can't connect to sandbox at localhost:${sandboxPort}")
    }
  }

  /**
    * Start DAML Sandbox, initialize the client and the adapter
    * @throws IOException
    */
  @throws[IOException]
  def startSandbox(): Unit = {
    sandbox = getSandboxRunner(testScenario).start
    waitForConnection(damlLedgerClient, waitTimeout)
    setupApplication.accept(damlLedgerClient)
    ledgerAdapter.start(parties)
  }

  /*
   * Stop DAML Sandbox, shutdown the client and the adapter
   */
  def stopSandbox(): Unit = {
    damlLedgerClient.close()
    if (sandbox != null) {
      sandbox.destroy()
      sandbox.waitFor()
    }
    sandbox = null
  }
}

File: src/main/java/com/digitalasset/testing/ledger/SandboxRunner.java
/*
 * Copyright 2019 Digital Asset (Switzerland) GmbH and/or its affiliates
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package com.digitalasset.testing.ledger;

import com.digitalasset.testing.junit4.LogLevel;
import java.io.File;
import java.io.IOException;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public abstract class SandboxRunner {
  private final Logger logger = LoggerFactory.getLogger(getClass());
  private final Path relativeDarPath;
  private final Integer sandboxPort;
  private final boolean useWallclockTime;
  private final Optional<String> ledgerId;
  private final Optional<LogLevel> logLevel;
  private final File damlRoot;
  private Process sandbox;

  SandboxRunner(
      Path damlRoot,
      Path relativeDarPath,
      Integer sandboxPort,
      boolean useWallclockTime,
      Optional<String> ledgerId,
      Optional<LogLevel> logLevel) {
    this.damlRoot = damlRoot != null ? damlRoot.toFile() : null;
    this.relativeDarPath = relativeDarPath;
    this.sandboxPort = sandboxPort;
    this.useWallclockTime = useWallclockTime;
    this.ledgerId = ledgerId;
    this.logLevel = logLevel;
  }

  private List<String> getDamlSandboxStarterCommand() {
    List<String> commands = new ArrayList<>();
    commands.add(getDamlCommand());
    commands.add("sandbox");
    commands.add("--");
    addCustomCommands(commands);
    commands.add("-p");
    commands.add(sandboxPort.toString());
    commands.add(useWallclockTime ? "-w" : "-s");
    ledgerId.ifPresent(
        value -> {
          commands.add("--ledgerid");
          commands.add(value);
        });
    logLevel.ifPresent(
        value -> {
          commands.add("--log-level");
          commands.add(value.toString());
        });
    commands.add(relativeDarPath.toString());
    return commands;
  }

  public final void startSandbox() throws IOException {
    ProcessBuilder procBuilder =
        new ProcessBuilder(getDamlSandboxStarterCommand()).directory(damlRoot);

    ProcessBuilder.Redirect redirect =
        ProcessBuilder.Redirect.appendTo(new File("integration-test-sandbox.log"));
    logger.debug("Executing: {}", String.join(" ", procBuilder.command()));
    sandbox = procBuilder.redirectError(redirect).redirectOutput(redirect).start();
    logger.info("Starting sandbox...");
  }

  protected void addCustomCommands(List<String> commands) {}

  protected abstract String getDamlCommand();

  public final void stopSandbox() throws Exception {
    if (sandbox != null) {
      logger.info("Stopping sandbox...");
      closeSandbox(sandbox);
      sandbox.waitFor();
      logger.info("Stopped sandbox");
    }
    sandbox = null;
  }

  protected abstract void closeSandbox(Process sandbox) throws IOException;
}
